File: chess/pieces/knight.py

from chess import PieceColorEnum
from chess.pieces.default_piece import DefaultPiece
from chess.pieces.piece import BoardSize, Piece, PiecePosition


class Knight(DefaultPiece):
    def __init__(self, color: PieceColorEnum, board_size: BoardSize, position: PiecePosition) -> None:
        super().__init__(color, board_size, position)

    def update_movements(self, board: list[list[Piece|None]]) -> None:
        self.clean_movements()

        FORWARD_DISTANCE = 2
        SIDE_STEP_DISTANCE = 1

        directions: list[PiecePosition] = [{'x': 0, 'y': -1},
                                           {'x': 1, 'y': 0},
                                           {'x': 0, 'y': 1},
                                           {'x': -1, 'y': 0}]

        for direction in directions:
            for sidestep_weight in [1, -1]:
                x = self.position['x'] + direction['x'] * FORWARD_DISTANCE + direction['y'] * SIDE_STEP_DISTANCE * sidestep_weight
                y = self.position['y'] + direction['y'] * FORWARD_DISTANCE + direction['x'] * SIDE_STEP_DISTANCE * sidestep_weight

                if not 0 <= x < self.board_size['width'] or not 0 <= y < self.board_size['height']:
                    continue

                target = board[y][x]

                if target == None:
                    self.movements[y][x] = True
                    continue

                if target.color != self.color:
                    self.movements[y][x] = True

File: src/main/java/com/vini/app/pieces/Knight.java
package com.vini.app.pieces;

import java.util.List;

public class Knight extends Piece {
	private final int[][] directions = {{0, -1}, {1, 0}, {0, 1}, {-1, 0}};
	private final int forwardStepWight = 2;
	private final int sideStepWeight = 1;

	@Override
	public IPiece updateMoves(List<List<IPiece>> board) {
		for (int[] direction : this.directions) {
			for (int sideStepDirection : new int[]{-1, 1}) {
				int[] position = {
					this.position()[0] + this.forwardStepWight * direction[1],
					this.position()[1] + this.forwardStepWight * direction[0]
				};

				position[0] = Math.abs(direction[1]) * sideStepDirection * sideStepWeight;
				position[1] = Math.abs(direction[0]) * sideStepDirection * sideStepWeight;

				if (!this.insideBoard(board, position)) {
					continue;
				}

				IPiece target = board.get(position()[1]).get(position()[0]);

				if (this.isFriend(this, target)) {
					continue;
				}

				this.moves().get(position[1]).set(position[0], true);
			}
		}

		return this;
	}
}

